src/XAD/Hessian.hpp

```diff
@@ -0,0 +1,152 @@
+/*******************************************************************************
+
+   Implementation of hessian computing methods.
+
+   This file is part of XAD, a comprehensive C++ library for
+   automatic differentiation.
+
+   Copyright (C) 2010-2024 Xcelerit Computing Ltd.
+
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU Affero General Public License as published
+   by the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU Affero General Public License for more details.
+
+   You should have received a copy of the GNU Affero General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+******************************************************************************/
+
+#include <XAD/TypeTraits.hpp>
+#include <XAD/XAD.hpp>
+
+#include <functional>
+#include <type_traits>
+#include <vector>
+
+namespace xad
+{
+
+// fwd_adj 2d vector
+template <typename T>
+std::vector<std::vector<T>> computeHessian(
+    const std::vector<xad::AReal<xad::FReal<T>>> &vec,
+    std::function<xad::AReal<xad::FReal<T>>(std::vector<xad::AReal<xad::FReal<T>>> &)> foo,
+    xad::Tape<xad::FReal<T>> *tape = xad::Tape<xad::FReal<T>>::getActive())
+{
+    std::vector<std::vector<T>> matrix(vec.size(), std::vector<T>(vec.size(), 0.0));
+    computeHessian(vec, foo, begin(matrix), end(matrix), tape);
+    return matrix;
+}
+
+// fwd_adj iterator
+template <class RowIterator, typename T>
+void computeHessian(
+    const std::vector<xad::AReal<xad::FReal<T>>> &vec,
+    std::function<xad::AReal<xad::FReal<T>>(std::vector<xad::AReal<xad::FReal<T>>> &)> foo,
+    RowIterator first, RowIterator last,
+    xad::Tape<xad::FReal<T>> *tape = xad::Tape<xad::FReal<T>>::getActive())
+{
+    auto v(vec);
+    unsigned int domain(static_cast<unsigned int>(v.size()));
+
+    if (std::distance(first, last) != domain)
+        throw OutOfRange("Iterator not allocated enough space");
+    static_assert(
+        !xad::detail::has_begin<typename std::iterator_traits<RowIterator>::value_type>::value,
```

Comment:
Why is this `!has_begin`? It looks the wrong way around - and I wonder why this goes through?

Reply:
I actually swapped the return types at some point but I reverted that now, and so the inverter is no longer needed. Could you look at the tests for has_begin for me? I'm not fully sure why an iterator type doesn't have a begin() method.